File: flutter_readium_platform_interface/lib/src/shared/guided_navigation/guided_navigation_description.dart
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';

import '../../utils/index.dart';
import 'guided_navigation_text.dart';

/// Alternative description for a guided navigation object, using one or more
/// media references.
///
/// At least one of [audioref], [imgref], [textref], [videoref], or [text]
/// must be present.
///
/// See https://readium.org/guided-navigation/schema/description.schema.json
@immutable
class GuidedNavigationDescription with EquatableMixin implements JSONable {
  const GuidedNavigationDescription({this.audioref, this.imgref, this.textref, this.videoref, this.text});

  final String? audioref;
  final String? imgref;
  final String? textref;
  final String? videoref;
  final GuidedNavigationText? text;

  @override
  List<Object?> get props => [audioref, imgref, textref, videoref, text];

  @override
  Map<String, dynamic> toJson() => {}
    ..putOpt('audioref', audioref)
    ..putOpt('imgref', imgref)
    ..putOpt('textref', textref)
    ..putOpt('videoref', videoref)
    ..putJSONableIfNotEmpty('text', text);

  static GuidedNavigationDescription? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;

    final jsonObject = Map<String, dynamic>.of(json);
    final audioref = jsonObject.optNullableString('audioref', remove: true);
    final imgref = jsonObject.optNullableString('imgref', remove: true);
    final textref = jsonObject.optNullableString('textref', remove: true);
    final videoref = jsonObject.optNullableString('videoref', remove: true);
    final text = GuidedNavigationText.fromJson(jsonObject.opt('text', remove: true));

    if (audioref == null && imgref == null && textref == null && videoref == null && text == null) {
      ReadiumLog.d('GuidedNavigationDescription: at least one media reference is required');
      return null;
    }

    return GuidedNavigationDescription(
      audioref: audioref,
      imgref: imgref,
      textref: textref,
      videoref: videoref,
      text: text,
    );
  }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/guided-nav/GuidedNavigationDescription.swift
import Foundation

/// Alternative description for a guided navigation object, using one or more
/// media references.
///
/// At least one of audioref, imgref, textref, videoref, or text must be present.
///
/// See https://readium.org/guided-navigation/schema/description.schema.json
struct GuidedNavigationDescription: Equatable {
  let audioref: String?
  let imgref: String?
  let textref: String?
  let videoref: String?
  let text: GuidedNavigationText?

  func toJson() -> [String: Any] {
    var res: [String: Any] = [:]
    if let audioref = audioref { res["audioref"] = audioref }
    if let imgref = imgref { res["imgref"] = imgref }
    if let textref = textref { res["textref"] = textref }
    if let videoref = videoref { res["videoref"] = videoref }
    if let text = text { res["text"] = text.toJson() }
    return res
  }

  static func fromJson(_ json: [String: Any]?) -> GuidedNavigationDescription? {
    guard let json = json else { return nil }

    let audioref = json["audioref"] as? String
    let imgref = json["imgref"] as? String
    let textref = json["textref"] as? String
    let videoref = json["videoref"] as? String
    let text = GuidedNavigationText.fromJson(json["text"])

    guard audioref != nil || imgref != nil || textref != nil
            || videoref != nil || text != nil else {
      return nil
    }

    return GuidedNavigationDescription(
      audioref: audioref,
      imgref: imgref,
      textref: textref,
      videoref: videoref,
      text: text
    )
  }
}
